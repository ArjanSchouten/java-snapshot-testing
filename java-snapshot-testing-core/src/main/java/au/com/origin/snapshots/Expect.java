package au.com.origin.snapshots;

import au.com.origin.snapshots.comparators.SnapshotComparator;
import au.com.origin.snapshots.reporters.SnapshotReporter;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

@RequiredArgsConstructor
public class Expect {
  private final SnapshotVerifier snapshotVerifier;
  private final Method testMethod;

  private SnapshotSerializer snapshotSerializer;
  private SnapshotComparator snapshotComparator;
  private List<SnapshotReporter> snapshotReporters;
  private String scenario;

  public static Expect of(SnapshotVerifier snapshotVerifier, Method method) {
    return new Expect(snapshotVerifier, method);
  }

  /**
   * Make an assertion on the given input parameters against what already exists
   *
   * @param firstObject first snapshot object
   * @param objects     other snapshot objects
   */
  public void toMatchSnapshot(Object firstObject, Object... objects) {
    Snapshot snapshot = snapshotVerifier.expectCondition(testMethod, firstObject, objects);
    if (snapshotSerializer != null) {
      snapshot.setSnapshotSerializer(snapshotSerializer);
    }
    if (snapshotComparator != null) {
      snapshot.setSnapshotComparator(snapshotComparator);
    }
    if (snapshotReporters != null) {
      snapshot.setSnapshotReporters(snapshotReporters);
    }
    if (scenario != null) {
      snapshot.setScenario(scenario);
    }
    snapshot.toMatchSnapshot();
  }


  /**
   * Normally a snapshot can be applied only once to a test method.
   * <p>
   * For Parameterized tests where the same method is executed multiple times you can supply
   * the scenario() to overcome this restriction.  Ensure each scenario is unique.
   *
   * @param scenario - unique scenario description
   * @return Snapshot
   */
  public Expect scenario(String scenario) {
    this.scenario = scenario;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param serializer your custom serializer
   * @return Snapshot
   */
  public Expect serializer(SnapshotSerializer serializer) {
    this.snapshotSerializer = serializer;
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot
   *
   * @param name - the {name} attribute serializer.{name} from snapshot.properties
   * @return Snapshot
   */
  public Expect serializer(String name) {
    this.snapshotSerializer = SnapshotProperties.getInstance("serializer." + name);
    return this;
  }

  /**
   * Apply a custom comparator for this snapshot
   *
   * @param comparator your custom comparator
   * @return Snapshot
   */
  public Expect comparator(SnapshotComparator comparator) {
    this.snapshotComparator = comparator;
    return this;
  }

  /**
   * Apply a list of custom reporters for this snapshot
   * This will replace the default reporters defined in the config
   *
   * @param reporters your custom reporters
   * @return Snapshot
   */
  public Expect reporters(SnapshotReporter... reporters) {
    this.snapshotReporters = Arrays.asList(reporters);
    return this;
  }

  /**
   * Apply a custom serializer for this snapshot.
   *
   * @param serializer your custom serializer
   * @return this
   * @see au.com.origin.snapshots.serializers.SnapshotSerializer
   * <p>
   * Example implementations
   * @see au.com.origin.snapshots.serializers.ToStringSnapshotSerializer
   * @see au.com.origin.snapshots.serializers.Base64SnapshotSerializer
   */
  @SneakyThrows
  public Expect serializer(Class<? extends SnapshotSerializer> serializer) {
    this.snapshotSerializer = serializer.getConstructor().newInstance();
    return this;
  }

}
