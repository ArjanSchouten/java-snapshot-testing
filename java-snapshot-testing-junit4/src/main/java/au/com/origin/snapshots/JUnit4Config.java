package au.com.origin.snapshots;

import java.lang.reflect.Method;
import java.util.stream.Stream;


public class JUnit4Config implements SnapshotConfig {

  @Override
  public StackTraceElement findStacktraceElement() {
    StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
    int elementsToSkip = 1; // Start after stackTrace
    while (SnapshotMatcher.class
        .getName()
        .equals(stackTraceElements[elementsToSkip].getClassName())) {
      elementsToSkip++;
    }

    return Stream.of(stackTraceElements)
        .skip(elementsToSkip)
        .filter(
            stackTraceElement ->
                hasTestAnnotation(
                    ReflectionUtilities.getMethod(
                        getClassForName(stackTraceElement.getClassName()),
                        stackTraceElement.getMethodName())))
        .findFirst()
        .orElseThrow(
            () ->
                new SnapshotMatchException(
                    "Could not locate a method with one of supported test annotations"));
  }

  protected boolean hasTestAnnotation(Method method) {
    return
    // Junit 4
    method.isAnnotationPresent(org.junit.Test.class)
        || method.isAnnotationPresent(org.junit.BeforeClass.class);
  }

  private Class<?> getClassForName(String className) {
    try {
      return Class.forName(className);
    } catch (ClassNotFoundException e) {
      throw new IllegalArgumentException(e);
    }
  }
}
