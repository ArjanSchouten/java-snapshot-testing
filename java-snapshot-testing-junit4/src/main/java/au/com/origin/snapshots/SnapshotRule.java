package au.com.origin.snapshots;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class SnapshotRule implements TestRule {
    @Override
    public Statement apply(Statement base, Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                SnapshotMatcher.start(new JUnit4Config(), description.getTestClass());
                try {
                    base.evaluate();
                } finally {
                    SnapshotMatcher.validateSnapshots();
                }
            }
        };
    }
}
