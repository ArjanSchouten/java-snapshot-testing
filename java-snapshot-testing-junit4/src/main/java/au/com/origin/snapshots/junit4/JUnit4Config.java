package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotConfig;
import org.apache.commons.lang3.NotImplementedException;

import java.lang.reflect.Method;

public class JUnit4Config implements SnapshotConfig {

  @Override
  public Class<?> getTestClass() {
    throw new NotImplementedException("You forgot to implement the @ClassRule for SnapshotClassRule");
  }

  @Override
  public Method getTestMethod(Class<?> testClass) {
    throw new NotImplementedException("You forgot to implement the @Rule for SnapshotRule");
  }

}
