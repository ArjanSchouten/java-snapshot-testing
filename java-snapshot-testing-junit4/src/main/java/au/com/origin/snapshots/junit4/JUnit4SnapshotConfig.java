package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotConfig;

import java.lang.reflect.Method;

public class JUnit4SnapshotConfig implements SnapshotConfig {

  @Override
  public Class<?> getTestClass() {
    throw new RuntimeException("You forgot to implement the @ClassRule for SnapshotClassRule");
  }

  @Override
  public Method getTestMethod(Class<?> testClass) {
    throw new RuntimeException("You forgot to implement the @Rule for SnapshotRule");
  }

}
