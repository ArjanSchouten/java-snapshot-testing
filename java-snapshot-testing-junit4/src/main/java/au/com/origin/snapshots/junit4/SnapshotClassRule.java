package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotConfig;
import au.com.origin.snapshots.SnapshotConfigInjector;
import au.com.origin.snapshots.SnapshotMatcher;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class SnapshotClassRule implements TestRule, SnapshotConfigInjector {
    @Override
    public Statement apply(Statement base, Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                // don't fail if a test is run alone from the IDE for example
                boolean failOnOrphans = description.getChildren().size() > 1;
                SnapshotMatcher.start(getSnapshotConfig(), failOnOrphans, description.getTestClass());
                try {
                    base.evaluate();
                } finally {
                    SnapshotMatcher.validateSnapshots();
                }
            }
        };
    }

    @Override
    public SnapshotConfig getSnapshotConfig() {
        return new JUnit4SnapshotConfig();
    }
}
