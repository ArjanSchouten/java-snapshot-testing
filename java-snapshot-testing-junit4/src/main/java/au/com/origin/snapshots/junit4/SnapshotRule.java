package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotMatcher;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class SnapshotRule implements TestRule {
    @Override
    public Statement apply(Statement base, Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                SnapshotMatcher.setTestMethod(description.getTestClass().getMethod(description.getMethodName()));
                base.evaluate();
            }
        };
    }
}
