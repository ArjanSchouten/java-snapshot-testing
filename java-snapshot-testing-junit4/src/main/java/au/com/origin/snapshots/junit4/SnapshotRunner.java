package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.*;
import org.junit.Test;
import org.junit.runner.notification.RunNotifier;
import org.junit.runners.BlockJUnit4ClassRunner;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.InitializationError;
import org.junit.runners.model.Statement;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

/**
 * Loosely based on: https://stackoverflow.com/questions/27745691/how-to-combine-runwith-with-runwithparameterized-class
 * <p>
 * This implementation is restricted because it appears to not support Parameterized tests
 */
public class SnapshotRunner extends BlockJUnit4ClassRunner implements SnapshotConfigInjector {

  Object testInstance;
  SnapshotVerifier snapshotVerifier;

  public SnapshotRunner(Class<?> klass) throws InitializationError {
    super(klass);
  }

  @Override
  protected Statement methodInvoker(FrameworkMethod method, Object test) {
    boolean isTest = method.getMethod().isAnnotationPresent(Test.class);
    if (isTest) {
      updateInstanceVariable(method.getMethod());
      boolean shouldInjectMethodArgument = Arrays.asList(method.getMethod().getParameterTypes()).contains(Expect.class);
      if (shouldInjectMethodArgument) {
        return new Statement() {
          @Override
          public void evaluate() throws Throwable {
            method.invokeExplosively(test, new Expect(snapshotVerifier, method.getMethod()));
          }
        };
      }
    }

    return super.methodInvoker(method, test);
  }

  @Override
  protected Object createTest() throws Exception {
    testInstance = super.createTest();
    return testInstance;
  }

  private void updateInstanceVariable(Method testMethod) {
    Arrays.stream(testInstance.getClass().getDeclaredFields())
        .filter(it -> it.getType() == Expect.class)
        .findFirst()
        .ifPresent(field -> {
          Expect expect = Expect.of(snapshotVerifier, testMethod);
          field.setAccessible(true);
          try {
            field.set(testInstance, expect);
          } catch (IllegalAccessException e) {
            throw new RuntimeException(e);
          }
        });
  }

  @Override
  public void run(RunNotifier notifier) {
    // We don't want the orphan check to happen when the user runs a single test in their IDE
    boolean failOnOrphans = getDescription().getChildren().size() > 1;
    snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), getTestClass().getJavaClass(), failOnOrphans);
    super.run(notifier);
    snapshotVerifier.validateSnapshots();
  }

  @Override
  protected void validateTestMethods(List<Throwable> errors) {
    // Disable as it checks for zero arguments
  }

  @Override
  public SnapshotConfig getSnapshotConfig() {
    return new PropertyResolvingSnapshotConfig();
  }

}

