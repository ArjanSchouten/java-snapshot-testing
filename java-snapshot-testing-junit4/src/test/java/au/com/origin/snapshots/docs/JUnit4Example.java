package au.com.origin.snapshots.docs;

import au.com.origin.snapshots.junit4.SnapshotRunner;
import au.com.origin.snapshots.Expect;
import org.junit.Test;
import org.junit.runner.RunWith;

// Ensure you RunWith the SnapshotRunner
@RunWith(SnapshotRunner.class)
public class JUnit4Example {

  @Test
  public void myTest1(Expect expect) {
    // Verify your snapshot
    expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void myTest2(Expect expect) {
    expect.toMatchSnapshot("Hello World Again");
  }
}
