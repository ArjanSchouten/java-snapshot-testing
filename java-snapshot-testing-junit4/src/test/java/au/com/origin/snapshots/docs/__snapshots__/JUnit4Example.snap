au.com.origin.snapshots.docs.JUnit4Example.myTest1=[
Hello World
]


au.com.origin.snapshots.docs.JUnit4Example.myTest2=[
Hello World Again
]